crates/sui-json-rpc/src/name_service.rs

```diff
@@ -263,14 +290,11 @@ fn validate_label(label: &str) -> Result<&str, NameServiceError> {
 
 impl fmt::Display for Domain {
```

Comment:
Rust already has a notion of ["alternate" formats](https://doc.rust-lang.org/std/fmt/#sign0) which you could leverage in your implementation. Doing it this way (making the `Display` impl aware of the alternate format) would allow you to avoid the pattern of creating a string in order to output it.

You can figure out whether someone has requested the alternate display format using the `.alternate()` flag on the `f: &mut fmt::Formatter<'_>`, and output using the new separator in that case:

```rust
impl fmt::Display for Domain {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut sep = '';
        let mut prefix = self.labels.len();

        // SAFETY: labels vector is guaranteed to contain at least 2 elements.
        if f.alternate() {
          prefix -= 2;
        }

        for label in &self.labels[0..prefix] {
          f.write_char(sep)?;
          f.write_str(label)?;
          sep = ACCEPTED_SEPARATORS[0];
        }
    
        if f.alternate() {
          f.write_char(SUI_NEW_FORMAT_SEPARATOR);
          f.write_str(self.labels[prefix]);
        }
    }
}
```

- You can get the default format with `format!("{domain}")`, `format!("{}", domain)` or `domain.to_string()`, and 
- you can get the default format using `format!("{domain:#}")`, or `format!("{:#}", domain)`.

Then at the call site, you can do something like this:

```rust
if let Some(DomainFormat::At) = format {
  format!("{d:#}")
} else {
  format!("{d}")
}
```

This technique only works if you expect there are only one of two alternates ever. Otherwise, the recommended approach is to have a `display` function that wraps your type (a reference of it) and implements your alternate display formats. This is what we do for displaying types canonically:

https://github.com/MystenLabs/sui/blob/b9751262bcb15b4147c234fb1a4875dd9fe91f02/external-crates/move/crates/move-core-types/src/language_storage.rs#L78-L108

Reply:
Thanks for the context @amnn! 

Will look into these/update the approach in a follow-up PR.